broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,399 @@
+	Lookup(uint64) (BacklogSegment, error)
+}
+
+// backlog stores backlogSegments and provides the ability to read/write
+// messages.
+type backlog struct {
+	head          atomic.Pointer[backlogSegment]
+	tail          atomic.Pointer[backlogSegment]
+	lookupLock    sync.RWMutex
+	lookupByIndex map[uint64]*backlogSegment
+	config        ConfigFetcher
+	messageCount  atomic.Uint64
+}
+
+// NewBacklog creates a backlog.
+func NewBacklog(c ConfigFetcher) Backlog {
+	lookup := make(map[uint64]*backlogSegment)
+	return &backlog{
+		lookupByIndex: lookup,
+		config:        c,
+	}
+}
+
+// Head return the head backlogSegment within the backlog.
+func (b *backlog) Head() BacklogSegment {
+	return b.head.Load()
+}
+
+// Append will add the given messages to the backlogSegment at head until
+// that segment reaches its limit. If messages remain to be added a new segment
+// will be created.
+func (b *backlog) Append(bm *m.BroadcastMessage) error {
+
+	if bm.ConfirmedSequenceNumberMessage != nil {
+		b.delete(uint64(bm.ConfirmedSequenceNumberMessage.SequenceNumber))
+		// TODO(clamb): add to metric?
+	}
+
+	for _, msg := range bm.Messages {
+		segment := b.tail.Load()
+		if segment == nil {
+			segment = newBacklogSegment()
+			b.head.Store(segment)
+			b.tail.Store(segment)
+		}
+
+		prevMsgIdx := segment.End()
+		if segment.count() >= b.config().SegmentLimit {
+			nextSegment := newBacklogSegment()
+			segment.nextSegment.Store(nextSegment)
+			prevMsgIdx = segment.End()
+			nextSegment.previousSegment.Store(segment)
+			segment = nextSegment
+			b.tail.Store(segment)
+		}
+
+		err := segment.append(prevMsgIdx, msg)
+		if errors.Is(err, errDropSegments) {
+			head := b.head.Load()
+			b.removeFromLookup(head.Start(), uint64(msg.SequenceNumber))
+			b.head.Store(segment)
+			b.tail.Store(segment)
+			b.messageCount.Store(0)
+			log.Warn(err.Error())
+		} else if errors.Is(err, errSequenceNumberSeen) {
+			log.Info("ignoring message sequence number (%s), already in backlog", msg.SequenceNumber)
+			continue
+		} else if err != nil {
+			return err
+		}
+		b.lookupLock.Lock()
+		b.lookupByIndex[uint64(msg.SequenceNumber)] = segment
+		b.lookupLock.Unlock()
+		b.messageCount.Add(1)
+	}
+
+	return nil
+}
+
+// Get reads messages from the given start to end MessageIndex.
+func (b *backlog) Get(start, end uint64) (*m.BroadcastMessage, error) {
+	head := b.head.Load()
+	tail := b.tail.Load()
+	if head == nil && tail == nil {
+		return nil, errOutOfBounds
+	}
+
+	if start < head.Start() {
+		start = head.Start()
+	}
+
+	if end > tail.End() {
+		return nil, errOutOfBounds
+	}
+
+	found, err := b.Lookup(start)
+	if err != nil {
+		return nil, err
+	}
+	segment, ok := found.(*backlogSegment)
+	if !ok {
+		return nil, fmt.Errorf("error in backlogSegment type assertion")
+	}
+
+	bm := &m.BroadcastMessage{Version: 1}
+	required := int(end-start) + 1
+	for {
+		segMsgs, err := segment.get(arbmath.MaxInt(start, segment.Start()), arbmath.MinInt(end, segment.End()))
```

Comment:
Is the MaxInt required here? Shouldn't `segment` always contain `start`?

Reply:
I think that the `MaxInt` is required here. The `start` given to this method could be for a message in any part of the backlog (in any segment) not necessarily the segment the for loop is currently looking at, hence the `MaxInt` with the start of the segment. For example there could be a backlog with segment length of 3 and the following messages [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]. If we did a `backlog.Get(1, 4)` it would need to look at messages over the first two segments. Once it got to the second segment it would be calling with `start = 1` but the second segment would not have message 1.